google/oauth2/service_account.py

```diff
@@ -451,6 +462,57 @@ def refresh(self, request):
             self.token = access_token
             self.expiry = expiry
 
+    def refresh(self, request):
+        """Refreshes the credential's access token.
+
+        This method is overridden to provide special handling for credentials that
+        use a self-signed JWT and have a trust boundary configured. In this
+        scenario, it first generates a temporary, IAM-specific self-signed JWT
+        to perform the trust boundary lookup, and then generates the final
+        self-signed JWT for the target API.
+
+        For all other cases, it falls back to the standard refresh behavior
+        from the parent class.
+
+        Args:
+            request (google.auth.transport.Request): The object used to make
+                HTTP requests.
+
+        Raises:
+            google.auth.exceptions.RefreshError: If the credentials could
+                not be refreshed.
+        """
+        # This is a special path for self-signed JWTs that need to look up a trust boundary.
+        # The `_subject` check is to ensure we are not in a domain-wide
+        # delegation flow, which uses a different authentication mechanism.
+        if (
+            self._always_use_jwt_access
+            and self._subject is None
+            and self._is_trust_boundary_lookup_required()
+        ):
+            # Special case: self-signed JWT with trust boundary.
+            # 1. Create a temporary self-signed JWT for the IAM API.
+            iam_audience = "https://iamcredentials.{}/".format(self._universe_domain)
+            iam_jwt_creds = jwt.Credentials.from_signing_credentials(self, iam_audience)
+            iam_jwt_creds.refresh(request)
+
+            # 2. Use this JWT to perform the trust boundary lookup.
+            # We temporarily set self.token for the base lookup method.
+            # The base lookup method will call self.apply() which adds the
+            # authorization header.
+            original_token = self.token
```

Comment:
This seems to be welcoming race conditions. What happens if one thread is updating trust boundary and set `self.token` to iam audience and another thread tries to read `self.token`

Reply:
Reverted this change since we're still waiting to hear back from the backend team on whether or not the workaround is necessary. We'll address self signed jwt flow if necessary in a future PR.